kubeadm: skip integration tests if kubeadm-cmd-skip flag passed

dmmcquay opened the pull request:
Will skip integration tests for token generation if it can't find a file by the given --kubeadm-path or default value.

**What this PR does / why we need it**: Tests would fail if just running `go test` in the dir because it expects to have more values. This won't change the behavior of `make test-cmd` which gets run here:
https://github.com/kubernetes/kubernetes/blob/master/Makefile#L258

**Which issue this PR fixes**: fixes #40155 

**Special notes for your reviewer**: /cc @pires @pipejakob @liggitt 

```release-note
NONE
```

dmmcquay committed:
kubeadm: skip tests if binary not present

Will skip integration tests for token generation if it can't find a file
by the given --kubeadm-path or default value.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40175)
<!-- Reviewable:end -->

liggitt commented on cmd/kubeadm/test/token_test.go:
```diff
@@ -29,12 +29,20 @@ const (
 
 func init() {
 	flag.StringVar(&kubeadmPath, "kubeadm-path", filepath.Join(os.Getenv("KUBE_ROOT"), "cluster/kubeadm.sh"), "Location of kubeadm")
+	if _, err := os.Stat(kubeadmPath); os.IsNotExist(err) {
```
this really should not be in an init() block... make it a helper that each test calls. global flag registration also does not belong in an init() block of a test file, but that can be a separate issue

dmmcquay replied:
I've seen this type of check happen in `init()` blocks before, especially when trying to test subprocesses:

https://speakerdeck.com/mitchellh/advanced-testing-with-go

dmmcquay replied:
I can move the flag registration out of there, shouldn't be a problem.

liggitt commented on cmd/kubeadm/test/token_test.go:
```diff
@@ -28,12 +28,11 @@ const (
 )
 
-var kubeadmPath string
+var kubeadmPath = flag.String("kubeadm-path", filepath.Join(os.Getenv("KUBE_ROOT"), "cluster/kubeadm.sh"), "Location of kubeadm")
```
sorry, the global package-level flag registration is the objection, not whether it's in or out of an init() block

dmmcquay replied:
I'm confused, are you suggesting not to have a package level flag? And if so, only use env variable to find kubeadm binary or helper `.sh`?

liggitt replied:
correct. as a rule, package-level unit tests should be self-contained, not depend on flags or external built binaries. 

exceptions live in our e2e and integration packages, not sprinkled through the cmd or pkg packages.

these two tests really seem like they should follow the test-cmd.sh pattern, building a binary and exercising invoking and testing the output via a shell script, rather than live as package unit tests

pipejakob replied:
They follow the test-cmd.sh pattern in most ways except for the fact that the tests themselves are written in Go instead of bash. As Derek pointed out, there was a lot of discussion around exactly where these tests should live and how they should work during the initial PR, and we came to this consensus then.

Also, in general the project's direction seems to favor replacing bash with Go code, so I'd prefer to keep these in Go.

dmmcquay commented:
@luxas Seems like the discussion has ended, want to take a look?

dmmcquay commented:
@luxas ping

luxas commented:
@liggitt do you like the idea of moving these tests to `test/integration/kubeadm` instead?

liggitt commented:
> @liggitt do you like the idea of moving these tests to test/integration/kubeadm instead?

yes, things that depend on external binaries and don't follow normal package unit test conventions (e.g. work with `go test ./foo/bar` without additional setup) should be under test/integration or test/e2e

luxas commented:
@dmmcquay There you go then. Both I and @liggitt are of the same opinion, please move this to a integration test package.

dmmcquay commented:
@luxas @liggitt I moved the tests to test/integration/kubeadm . Also moved the bazel `BUILD` file, although I am not sure that is needed. 

Please advise.

dmmcquay commented:
Looks like the bazel file was throwing it off. Removed for now.

dmmcquay commented:
@k8s-bot unit test this

dmmcquay commented:
@liggitt any ideas why this is failing unit/integration tests?

dmmcquay commented:
@liggitt I figured out what the problem was, tests should pass now.

dmmcquay commented:
@k8s-bot gci gce e2e test this

dmmcquay commented:
@luxas it passes all the tests, just needs a LGTM from you.

pires commented:
LGTM but please squash.

dmmcquay commented:
@k8s-bot cvm gce e2e test this

luxas commented on test/integration/kubeadm/token_test.go:
```diff
@@ -22,20 +22,29 @@ import (
 	"regexp"
 	"testing"
+
+	_ "github.com/golang/glog"
```
Why vendor this?

dmmcquay replied:
its required because of the flags that `make test-integration` pass to it. 

https://github.com/kubernetes/kubernetes/blob/master/hack/make-rules/test-integration.sh#L66

luxas reviewed (commented):
I had expected this file to start with `// +build integration`, I think that build flag should be present.

I'm now starting to think if these fit even better in on e2e tests... Maybe we should move them there instead? @liggitt @ixdy what do you think?

I just couldn't find any other exec-based tests than under `test/e2e` when I started looking into it.

dmmcquay commented:
@luxas I can add the build tag, that makes sense. My usage of Go build tags had been more OS specific, as I mentioned before, didn't know kubernetes had specific tags in place, which makes more sense.

dmmcquay commented:
I also wanted to point out that a discussion around where these tests should live has already taken place here: https://github.com/kubernetes/kubernetes/pull/35144

spxtr commented:
Note that if you run unit tests with bazel this isn't a problem, since the [BUILD](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/test/BUILD#L17-L28) file explicitly declares the dependency.

luxas commented:
@spxtr See https://docs.google.com/document/d/1d-xUyYe4DMjyV1-xB6xqn6FTiryN2FXnKa4JF1jKozc/edit#heading=h.fnh0z4ayjjjn (written by @dmmcquay) where we're discussing the e2e/integration and overall testing strategy for kubeadm.

mikedanese commented on test/integration/kubeadm/util.go:
```diff
@@ -1,3 +1,5 @@
+// +build integration
```
Generally I think we should discourage and stop using build tags like this. If we use build tags, we probably won't ever check that these tests actually compile.

mikedanese commented:
I don't mind the move, but skipping integration tests if binary is not present seems like a really good way to never run the tests. We should explicitly skip the tests, or build the binary.

dmmcquay commented:
@mikedanese thanks for taking a look. So if I got rid of the build tag and removed the skipping of the tests, you would approve of this move? 

Currently as this stands now, this test doesn't get run during `make test-integration`, but during `make test-cmd`. During `make test-integration` it gets skipped because it doesn't provide the `--kubeadm-path` flag ,where `make test-cmd` has the correct setup to do that.

dmmcquay commented:
@k8s-bot unit test this

pires commented:
Can you enable it on `make test-integration`?

spxtr commented:
I actually kind of like the test where it is now. Its dependency on the binary is cleanly expressed in the bazel rule. I'd rather not mangle `test/integration` any more than it currently is mangled.

liggitt commented:
> I actually kind of like the test where it is now. Its dependency on the binary is cleanly expressed in the bazel rule. I'd rather not mangle test/integration any more than it currently is mangled.

requiring a package to be compiled into an externally runnable binary in order to run package-level unit tests is... unexpected. apart from this test, standard go test commands (`go test ./cmd/kubeadm/...`) work perfectly

spxtr commented:
I agree it would be nice for the go tool to know what to do, but at the same time I greatly prefer this:
```
go_test(
    args = ["--kubeadm-path=../kubeadm"],
    data = ["//cmd/kubeadm"],
    (...)
)
```
over the bash that will be required to do this the other way.

luxas commented:
@spxtr Is it possible to make bazel first compile the binary before testing and then passing those args to the `_test.go` files in `cmd/kubeadm/test`? If that's possible, that'd be great of course!
I had no idea it was possible to do such things in bazel...

As mentioned above, in that case we're breaking the "normal" behaviour of the Go unit tests, but on the other hand I don't see any blocker to making `cmd/kubeadm/test` special **if we document it well enough.**

@spxtr Please join our sig call tomorrow, we're gonna discuss kubeadm testing: https://docs.google.com/document/d/1d-xUyYe4DMjyV1-xB6xqn6FTiryN2FXnKa4JF1jKozc/edit?usp=sharing

spxtr commented:
> @spxtr Is it possible to make bazel first compile the binary before testing and then passing those args to the _test.go files in cmd/kubeadm/test? If that's possible, that'd be great of course!
> I had no idea it was possible to do such things in bazel...

The line `data = ["//cmd/kubeadm"]` states that this test requires the output of the `//cmd/kubeadm` rule as input to the test. Bazel's build language makes this sort of dependency explicit, so if you say `bazel test //cmd/kubeadm/test/...` then it will notice that you need to build the binary before running these tests.

It also handles caching properly so you don't need to rebuild every time :)

dmmcquay committed:
kubeadm: updated cmd tests as discussed in SIG

After today's SIG meeting, it was discussed how to proceed with these
types of test-cmd tests.  They will live in kubeamd/test/cmd and will
provide a flag that will allow you to skip them (--kubeadm-cmd-skip) and
by default will fail if kubeadm binary is not present

dmmcquay commented:
@luxas This now contains the changes after debating in the SIG meeting today. PTAL so that https://github.com/kubernetes/kubernetes/pull/40045 can also be changed and merge in according to these changes.

spxtr commented:
Change it to `args = ["--kubeadm-path=../../kubeadm"]` in the BUILD file.

luxas commented:
/lgtm given tests pass

dmmcquay commented:
@luxas just needs your approval.

spxtr commented:
/approve

dmmcquay commented:
@jbeda can I get your approval on this PR?

luxas commented:
/approve
